File: minigit-4/py/commands/checkout.py
"""checkout command - Switch branches or restore files."""

import sys
import os
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from utils import find_repo_root, get_minigit_dir
from objects import read_object, Commit, Tree, resolve_object, hash_object
from index import Index, IndexEntry
from refs import (
    set_head, create_branch, get_ref, get_current_branch,
    resolve_head
)


def get_tree_files(tree_sha: str, repo_root: Path) -> dict:
    """Get all files from a tree as {path: (mode, sha)}."""
    files = {}

    def walk_tree(sha: str, prefix: str):
        try:
            _, data = read_object(sha, repo_root)
            tree = Tree.deserialize(data)
            for entry in tree.entries:
                path = f'{prefix}{entry.name}' if prefix else entry.name
                if entry.mode.startswith('40'):
                    walk_tree(entry.sha, path + '/')
                else:
                    files[path] = (entry.mode, entry.sha)
        except:
            pass

    walk_tree(tree_sha, '')
    return files


def run(args):
    """Switch branches or restore files."""
    repo_root = find_repo_root()
    if repo_root is None:
        print("fatal: not a minigit repository", file=sys.stderr)
        return 1

    # Parse arguments
    create_new = False
    target = None
    files = []
    separator_seen = False
    start_point = None

    i = 0
    while i < len(args):
        arg = args[i]
        if arg == '-b':
            create_new = True
        elif arg == '--':
            separator_seen = True
        elif separator_seen:
            files.append(arg)
        elif not arg.startswith('-'):
            if target is None:
                target = arg
            elif create_new and start_point is None:
                start_point = arg
            else:
                files.append(arg)
        i += 1

    if not target and not files:
        print("error: no target specified", file=sys.stderr)
        return 1

    # Handle file restore (checkout -- <files>)
    if files or separator_seen:
        return checkout_files(repo_root, target, files)

    # Handle branch checkout
    return checkout_ref(repo_root, target, create_new, start_point)


def checkout_files(repo_root: Path, source: str, files: list) -> int:
    """Restore files from index or commit."""
    index = Index.read(repo_root)

    if source:
        # Restore from specific commit
        try:
            sha = resolve_object(source, repo_root)
            _, data = read_object(sha, repo_root)
            commit = Commit.deserialize(data)
            tree_files = get_tree_files(commit.tree_sha, repo_root)
        except ValueError as e:
            print(f"fatal: {e}", file=sys.stderr)
            return 1
    else:
        # Restore from index
        tree_files = None

    for name in files:
        file_path = repo_root / name

        if tree_files is not None:
            # Restore from commit
            if name not in tree_files:
                print(f"error: pathspec '{name}' did not match any file(s)", file=sys.stderr)
                return 1

            mode, blob_sha = tree_files[name]
            _, content = read_object(blob_sha, repo_root)

            file_path.parent.mkdir(parents=True, exist_ok=True)
            file_path.write_bytes(content)

            # Update index
            entry = IndexEntry.from_file(file_path, blob_sha, repo_root)
            index.add_entry(entry)
        else:
            # Restore from index
            if name not in index.entries:
                print(f"error: pathspec '{name}' did not match any file(s)", file=sys.stderr)
                return 1

            entry = index.entries[name]
            _, content = read_object(entry.sha, repo_root)

            file_path.parent.mkdir(parents=True, exist_ok=True)
            file_path.write_bytes(content)

    index.write(repo_root)
    return 0


def checkout_ref(repo_root: Path, target: str, create_new: bool, start_point: str) -> int:
    """Checkout a branch or commit."""
    index = Index.read(repo_root)

    # Get current state
    current_head = resolve_head(repo_root)
    if current_head:
        _, data = read_object(current_head, repo_root)
        current_commit = Commit.deserialize(data)
        current_files = get_tree_files(current_commit.tree_sha, repo_root)
    else:
        current_files = {}

    # Resolve target
    if create_new:
        # Create new branch
        if start_point:
            try:
                sha = resolve_object(start_point, repo_root)
            except ValueError as e:
                print(f"fatal: {e}", file=sys.stderr)
                return 1
        else:
            sha = current_head
            if not sha:
                print("fatal: not a valid object name: 'HEAD'", file=sys.stderr)
                return 1

        if not create_branch(target, sha, repo_root):
            print(f"fatal: branch '{target}' already exists", file=sys.stderr)
            return 1

        set_head(f'refs/heads/{target}', symbolic=True, repo_root=repo_root)

        # Update working tree
        if sha != current_head:
            return update_worktree(repo_root, current_files, sha, index)

        return 0

    # Check if target is a branch
    branch_sha = get_ref(f'refs/heads/{target}', repo_root)
    is_branch = branch_sha is not None

    try:
        if is_branch:
            target_sha = branch_sha
        else:
            target_sha = resolve_object(target, repo_root)
    except ValueError as e:
        print(f"error: pathspec '{target}' did not match any file(s) known to minigit", file=sys.stderr)
        return 1

    # Check for uncommitted changes that would be overwritten
    _, target_data = read_object(target_sha, repo_root)
    target_commit = Commit.deserialize(target_data)
    target_files = get_tree_files(target_commit.tree_sha, repo_root)

    # Check for conflicts
    for name, (mode, sha) in target_files.items():
        if name in current_files:
            current_mode, current_sha = current_files[name]
            if current_sha != sha:
                # File differs - check working tree
                file_path = repo_root / name
                if file_path.exists():
                    try:
                        if file_path.is_symlink():
                            content = os.readlink(file_path).encode()
                        else:
                            content = file_path.read_bytes()
                        worktree_sha = hash_object(content, 'blob')
                        if worktree_sha != current_sha:
                            # Working tree has changes that would be overwritten
                            print(f"error: Your local changes to the following files would be overwritten by checkout:", file=sys.stderr)
                            print(f"\t{name}", file=sys.stderr)
                            print("Please commit your changes or stash them before you switch branches.", file=sys.stderr)
                            return 1
                    except:
                        pass

    # Update HEAD
    if is_branch:
        set_head(f'refs/heads/{target}', symbolic=True, repo_root=repo_root)
        branch_msg = f"Switched to branch '{target}'"
    else:
        set_head(target_sha, symbolic=False, repo_root=repo_root)
        branch_msg = f"HEAD is now at {target_sha[:7]}"

    # Update working tree
    result = update_worktree(repo_root, current_files, target_sha, index)
    if result == 0:
        print(branch_msg)
    return result


def update_worktree(repo_root: Path, current_files: dict, target_sha: str, index: Index) -> int:
    """Update working tree to match target commit."""
    _, target_data = read_object(target_sha, repo_root)
    target_commit = Commit.deserialize(target_data)
    target_files = get_tree_files(target_commit.tree_sha, repo_root)

    # Remove files not in target
    for name in current_files:
        if name not in target_files:
            file_path = repo_root / name
            if file_path.exists():
                file_path.unlink()
                # Remove empty directories
                try:
                    parent = file_path.parent
                    while parent != repo_root:
                        if not any(parent.iterdir()):
                            parent.rmdir()
                        parent = parent.parent
                except:
                    pass
            # Remove from index
            index.remove_entry(name)

    # Add/update files in target
    for name, (mode, sha) in target_files.items():
        file_path = repo_root / name
        _, content = read_object(sha, repo_root)

        file_path.parent.mkdir(parents=True, exist_ok=True)
        file_path.write_bytes(content)

        # Set executable bit if needed
        if mode == '100755':
            os.chmod(file_path, 0o755)

        # Update index
        entry = IndexEntry.from_file(file_path, sha, repo_root)
        index.add_entry(entry)

    index.write(repo_root)
    return 0

File: minigit-4/ts/src/commands/checkout.ts
// checkout command - Switch branches or restore files

import * as fs from 'fs';
import * as path from 'path';
import { findRepoRoot, normalizePathSeparator, ensureDir } from '../utils';
import { readIndex, writeIndex, createIndexEntryFromFile } from '../index-file';
import { readObject, parseCommitContent, parseTreeContent, createBlobContent, hashObject, objectExists } from '../objects';
import {
  resolveRevision,
  resolveRef,
  branchExists,
  updateBranch,
  writeHead,
  writeSymbolicRef,
  getHeadCommit,
  getCurrentBranch,
} from '../refs';
import { IndexEntry } from '../types';

export function checkout(args: string[]): number {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a minigit repository');
    return 1;
  }

  let createBranch = false;
  const positional: string[] = [];
  let inPathSpec = false;
  const pathSpecs: string[] = [];

  for (const arg of args) {
    if (arg === '-b') {
      createBranch = true;
    } else if (arg === '--') {
      inPathSpec = true;
    } else if (inPathSpec) {
      pathSpecs.push(arg);
    } else if (!arg.startsWith('-')) {
      positional.push(arg);
    }
  }

  // checkout -- <paths> : restore files from index
  if (pathSpecs.length > 0) {
    const commit = positional.length > 0 ? positional[0] : null;
    return restoreFiles(repoRoot, commit, pathSpecs);
  }

  // checkout -b <branch> [start-point]
  if (createBranch) {
    if (positional.length === 0) {
      console.error('fatal: branch name required');
      return 1;
    }
    const branchName = positional[0];
    const startPoint = positional[1];
    return createAndCheckoutBranch(repoRoot, branchName, startPoint);
  }

  // checkout <branch|commit>
  if (positional.length === 0) {
    console.error('fatal: checkout requires a branch name or commit');
    return 1;
  }

  const target = positional[0];

  // Check if it's a branch
  if (branchExists(repoRoot, target)) {
    return checkoutBranch(repoRoot, target);
  }

  // Try to resolve as commit
  const sha = resolveRevision(repoRoot, target);
  if (sha) {
    return checkoutDetached(repoRoot, sha);
  }

  console.error(`error: pathspec '${target}' did not match any file(s) known to git`);
  return 1;
}

function checkoutBranch(repoRoot: string, branchName: string): number {
  const sha = resolveRef(repoRoot, branchName);
  if (!sha) {
    console.error(`error: branch '${branchName}' not found`);
    return 1;
  }

  // Check for uncommitted changes that would be overwritten
  if (!canSafelyCheckout(repoRoot, sha)) {
    console.error('error: Your local changes to the following files would be overwritten by checkout');
    console.error('Please commit your changes or stash them before you switch branches.');
    return 1;
  }

  // Update working tree and index
  updateWorkingTree(repoRoot, sha);

  // Update HEAD
  writeSymbolicRef(repoRoot, 'HEAD', `refs/heads/${branchName}`);

  console.log(`Switched to branch '${branchName}'`);
  return 0;
}

function checkoutDetached(repoRoot: string, sha: string): number {
  // Check for uncommitted changes that would be overwritten
  if (!canSafelyCheckout(repoRoot, sha)) {
    console.error('error: Your local changes to the following files would be overwritten by checkout');
    console.error('Please commit your changes or stash them before you switch branches.');
    return 1;
  }

  // Update working tree and index
  updateWorkingTree(repoRoot, sha);

  // Update HEAD to detached state
  writeHead(repoRoot, sha);

  console.log(`HEAD is now at ${sha.slice(0, 7)}`);
  return 0;
}

function createAndCheckoutBranch(repoRoot: string, branchName: string, startPoint?: string): number {
  // Check if branch already exists
  if (branchExists(repoRoot, branchName)) {
    console.error(`fatal: branch '${branchName}' already exists`);
    return 1;
  }

  // Resolve start point
  let sha: string | null;
  if (startPoint) {
    sha = resolveRevision(repoRoot, startPoint);
    if (!sha) {
      console.error(`fatal: not a valid object name: '${startPoint}'`);
      return 1;
    }
  } else {
    sha = getHeadCommit(repoRoot);
    if (!sha) {
      console.error('fatal: not a valid object name: HEAD');
      return 1;
    }
  }

  // Check for uncommitted changes that would be overwritten
  if (!canSafelyCheckout(repoRoot, sha)) {
    console.error('error: Your local changes to the following files would be overwritten by checkout');
    return 1;
  }

  // Create branch
  updateBranch(repoRoot, branchName, sha);

  // Update working tree if start point is different from current
  const currentSha = getHeadCommit(repoRoot);
  if (currentSha !== sha) {
    updateWorkingTree(repoRoot, sha);
  }

  // Update HEAD
  writeSymbolicRef(repoRoot, 'HEAD', `refs/heads/${branchName}`);

  console.log(`Switched to a new branch '${branchName}'`);
  return 0;
}

function restoreFiles(repoRoot: string, commit: string | null, paths: string[]): number {
  let entries = readIndex(repoRoot);

  for (const pathSpec of paths) {
    const relativePath = normalizePathSeparator(pathSpec);

    if (commit) {
      // Restore from commit
      const sha = resolveRevision(repoRoot, commit);
      if (!sha) {
        console.error(`error: pathspec '${commit}' did not match any known ref`);
        return 1;
      }

      const { content } = readObject(repoRoot, sha);
      const commitInfo = parseCommitContent(content);
      const fileContent = getFileFromTree(repoRoot, commitInfo.tree, relativePath);

      if (fileContent === null) {
        console.error(`error: pathspec '${pathSpec}' did not match any file(s) known to git`);
        return 1;
      }

      // Write to working tree
      const fullPath = path.join(repoRoot, relativePath);
      ensureDir(path.dirname(fullPath));
      fs.writeFileSync(fullPath, fileContent);
    } else {
      // Restore from index
      const entry = entries.find((e) => e.name === relativePath);
      if (!entry) {
        console.error(`error: pathspec '${pathSpec}' did not match any file(s) known to git`);
        return 1;
      }

      const { content } = readObject(repoRoot, entry.sha);
      const fullPath = path.join(repoRoot, relativePath);
      ensureDir(path.dirname(fullPath));
      fs.writeFileSync(fullPath, content);
    }
  }

  return 0;
}

function canSafelyCheckout(repoRoot: string, targetSha: string): boolean {
  const entries = readIndex(repoRoot);
  const currentHead = getHeadCommit(repoRoot);

  // Get target tree files
  const targetFiles = new Map<string, string>();
  const { content } = readObject(repoRoot, targetSha);
  const commitInfo = parseCommitContent(content);
  collectTreeFiles(repoRoot, commitInfo.tree, '', targetFiles);

  // Get current tree files
  const currentFiles = new Map<string, string>();
  if (currentHead) {
    const { content: currentContent } = readObject(repoRoot, currentHead);
    const currentCommitInfo = parseCommitContent(currentContent);
    collectTreeFiles(repoRoot, currentCommitInfo.tree, '', currentFiles);
  }

  // Check for uncommitted changes that would be overwritten
  for (const entry of entries) {
    const fullPath = path.join(repoRoot, entry.name);

    if (!fs.existsSync(fullPath)) {
      // File is deleted in working tree
      continue;
    }

    const targetSha = targetFiles.get(entry.name);
    const currentSha = currentFiles.get(entry.name);

    // If target differs from current and working tree differs from index
    if (targetSha !== currentSha) {
      // Check if working tree matches index
      const stat = fs.lstatSync(fullPath);
      let content: Buffer;

      if (stat.isSymbolicLink()) {
        content = Buffer.from(fs.readlinkSync(fullPath));
      } else {
        content = fs.readFileSync(fullPath);
      }

      const blobContent = createBlobContent(content);
      const workingSha = hashObject(blobContent);

      if (workingSha !== entry.sha) {
        // Working tree has changes that would be lost
        return false;
      }
    }
  }

  return true;
}

function updateWorkingTree(repoRoot: string, sha: string): void {
  const entries = readIndex(repoRoot);

  // Get current tree files
  const currentHead = getHeadCommit(repoRoot);
  const currentFiles = new Set<string>();
  if (currentHead) {
    const { content } = readObject(repoRoot, currentHead);
    const commitInfo = parseCommitContent(content);
    collectTreeFilesSet(repoRoot, commitInfo.tree, '', currentFiles);
  }

  // Get target tree files
  const { content } = readObject(repoRoot, sha);
  const commitInfo = parseCommitContent(content);
  const targetFiles = new Map<string, { sha: string; mode: string }>();
  collectTreeFilesWithMode(repoRoot, commitInfo.tree, '', targetFiles);

  // Remove files that are in current but not in target
  for (const file of currentFiles) {
    if (!targetFiles.has(file)) {
      const fullPath = path.join(repoRoot, file);
      if (fs.existsSync(fullPath)) {
        fs.unlinkSync(fullPath);
        // Remove empty parent directories
        removeEmptyDirs(repoRoot, path.dirname(fullPath));
      }
    }
  }

  // Update/create files from target
  const newEntries: IndexEntry[] = [];

  for (const [name, { sha: fileSha, mode }] of targetFiles) {
    const fullPath = path.join(repoRoot, name);
    ensureDir(path.dirname(fullPath));

    const { content } = readObject(repoRoot, fileSha);

    if (mode === '120000') {
      // Symlink
      if (fs.existsSync(fullPath)) {
        fs.unlinkSync(fullPath);
      }
      fs.symlinkSync(content.toString(), fullPath);
    } else {
      fs.writeFileSync(fullPath, content);
      if (mode === '100755') {
        fs.chmodSync(fullPath, 0o755);
      }
    }

    // Create index entry
    const stat = fs.lstatSync(fullPath);
    newEntries.push({
      ctimeSec: Math.floor(stat.ctimeMs / 1000),
      ctimeNsec: Math.floor((stat.ctimeMs % 1000) * 1000000),
      mtimeSec: Math.floor(stat.mtimeMs / 1000),
      mtimeNsec: Math.floor((stat.mtimeMs % 1000) * 1000000),
      dev: stat.dev,
      ino: stat.ino,
      mode: parseInt(mode, 8),
      uid: stat.uid,
      gid: stat.gid,
      size: stat.size,
      sha: fileSha,
      flags: 0,
      name,
    });
  }

  writeIndex(repoRoot, newEntries);
}

function collectTreeFiles(repoRoot: string, treeSha: string, prefix: string, files: Map<string, string>): void {
  const { content } = readObject(repoRoot, treeSha);
  const entries = parseTreeContent(content);

  for (const entry of entries) {
    const fullName = prefix ? `${prefix}/${entry.name}` : entry.name;

    if (entry.mode === '40000') {
      collectTreeFiles(repoRoot, entry.sha, fullName, files);
    } else {
      files.set(fullName, entry.sha);
    }
  }
}

function collectTreeFilesSet(repoRoot: string, treeSha: string, prefix: string, files: Set<string>): void {
  const { content } = readObject(repoRoot, treeSha);
  const entries = parseTreeContent(content);

  for (const entry of entries) {
    const fullName = prefix ? `${prefix}/${entry.name}` : entry.name;

    if (entry.mode === '40000') {
      collectTreeFilesSet(repoRoot, entry.sha, fullName, files);
    } else {
      files.add(fullName);
    }
  }
}

function collectTreeFilesWithMode(
  repoRoot: string,
  treeSha: string,
  prefix: string,
  files: Map<string, { sha: string; mode: string }>
): void {
  const { content } = readObject(repoRoot, treeSha);
  const entries = parseTreeContent(content);

  for (const entry of entries) {
    const fullName = prefix ? `${prefix}/${entry.name}` : entry.name;

    if (entry.mode === '40000') {
      collectTreeFilesWithMode(repoRoot, entry.sha, fullName, files);
    } else {
      files.set(fullName, { sha: entry.sha, mode: entry.mode });
    }
  }
}

function getFileFromTree(repoRoot: string, treeSha: string, filePath: string): Buffer | null {
  const parts = filePath.split('/').filter((p) => p);
  let currentSha = treeSha;

  for (let i = 0; i < parts.length; i++) {
    const { content } = readObject(repoRoot, currentSha);
    const entries = parseTreeContent(content);
    const entry = entries.find((e) => e.name === parts[i]);

    if (!entry) {
      return null;
    }

    currentSha = entry.sha;
  }

  const { type, content } = readObject(repoRoot, currentSha);
  if (type !== 'blob') {
    return null;
  }

  return content;
}

function removeEmptyDirs(repoRoot: string, dirPath: string): void {
  while (dirPath !== repoRoot && dirPath.startsWith(repoRoot)) {
    try {
      const entries = fs.readdirSync(dirPath);
      if (entries.length === 0) {
        fs.rmdirSync(dirPath);
        dirPath = path.dirname(dirPath);
      } else {
        break;
      }
    } catch {
      break;
    }
  }
}
